pkg/broker/eventypes.go

```diff
@@ -0,0 +1,118 @@
+/*
+Copyright 2023 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package broker
+
+import (
+	"context"
+	"encoding/base64"
+	"github.com/cloudevents/sdk-go/v2/event"
+	"go.uber.org/zap"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	eventingv1 "knative.dev/eventing/pkg/apis/eventing/v1"
+	"knative.dev/eventing/pkg/apis/eventing/v1beta2"
+	"knative.dev/eventing/pkg/apis/feature"
+	eventingv1beta2 "knative.dev/eventing/pkg/client/clientset/versioned/typed/eventing/v1beta2"
+	eventinglisters "knative.dev/eventing/pkg/client/listers/eventing/v1"
+	v1beta22 "knative.dev/eventing/pkg/client/listers/eventing/v1beta2"
+	"knative.dev/eventing/pkg/utils"
+	"knative.dev/pkg/apis"
+	v1 "knative.dev/pkg/apis/duck/v1"
+)
+
+type EventtypeAutoHandler struct {
+	BrokerLister    eventinglisters.BrokerLister
+	EventTypeLister v1beta22.EventTypeLister
+	EventingClient  eventingv1beta2.EventingV1beta2Interface
+	FeatureStore    *feature.Store
+	Logger          *zap.Logger
+}
+
+func (h *EventtypeAutoHandler) getBroker(name, namespace string) (*eventingv1.Broker, error) {
+	broker, err := h.BrokerLister.Brokers(namespace).Get(name)
```

Comment:
in theory this is not limited to just brokers

Reply:
in `EventType` itself we just check if the thing exists:
https://github.com/knative/eventing/blob/main/pkg/reconciler/eventtype/eventtype.go#L44

So, perhaps some approach to make this as a broker independent library, that's used on "Reference Broker" and the Channels from core ( not all in one PR is needed).

Just giving some ideas

Reply:
I need to get both KRef and OwnerRef from the broker instance itself (namely uid). I was thinking that the signature of this functions might take this params as a input. That would make it reusable towards Channels.

Reply:
Agree David, the `AutoCreateEventType` caller knows the GVK, so the lib can accepts the GVK + coordinates (ns/name)